File: backend/src/aithru_agent/stream/writer.py
from datetime import UTC, datetime
from typing import TYPE_CHECKING, Any

from .events import AgentStreamEvent, AgentStreamRedaction, AgentStreamSource, AgentStreamVisibility
from .redaction import combine_redaction, redact_stream_payload

if TYPE_CHECKING:
    from aithru_agent.persistence.protocols import AgentEventStore


class AgentEventWriter:
    def __init__(self, store: "AgentEventStore") -> None:
        self._store = store

    async def write(
        self,
        *,
        run_id: str,
        type: str,
        source: AgentStreamSource | dict[str, str | None],
        payload: Any,
        thread_id: str | None = None,
        visibility: AgentStreamVisibility | str = AgentStreamVisibility.USER,
        redaction: AgentStreamRedaction | str = AgentStreamRedaction.NONE,
        summary: str | None = None,
    ) -> AgentStreamEvent:
        sequence = await self._store.next_sequence(run_id)
        safe_payload, detected_redaction = redact_stream_payload(payload)
        event = AgentStreamEvent(
            id=f"{run_id}:{sequence}",
            run_id=run_id,
            thread_id=thread_id,
            sequence=sequence,
            timestamp=datetime.now(UTC).isoformat().replace("+00:00", "Z"),
            type=type,
            source=source,
            visibility=visibility,
            redaction=combine_redaction(redaction, detected_redaction),
            summary=summary,
            payload=safe_payload,
        )
        await self._store.append(event)
        return event

File: backend/packages/stream/src/writer.ts
import { nanoid } from "nanoid";
import type { AgentStreamEvent, AgentStreamSource } from "@aithru-agent/contracts";
import { VISIBILITY, REDACTION } from "./events.js";
import { redactPayload } from "./redaction.js";

export interface AgentEventStore {
  appendEvent(runId: string, event: AgentStreamEvent): void;
  listEvents(runId: string): AgentStreamEvent[];
  nextEventSequence?(runId: string): number;
}

function generateEventId(): string {
  return `evt_${nanoid(12)}`;
}

export class AgentEventWriter {
  constructor(private store: AgentEventStore) {}

  write(
    runId: string,
    threadId: string | null,
    type: string,
    payload: unknown,
    opts?: {
      source?: AgentStreamSource;
      visibility?: typeof VISIBILITY[keyof typeof VISIBILITY];
      redaction?: typeof REDACTION[keyof typeof REDACTION];
      summary?: string;
    },
  ): AgentStreamEvent {
    const sequence = this.store.nextEventSequence?.(runId) ?? this.store.listEvents(runId).length + 1;
    const event: AgentStreamEvent = {
      id: generateEventId(),
      run_id: runId,
      thread_id: threadId,
      sequence,
      timestamp: new Date().toISOString().replace(/\.\d{3}/, ""),
      type,
      source: opts?.source || { kind: "system" },
      visibility: opts?.visibility || VISIBILITY.USER,
      redaction: opts?.redaction || REDACTION.NONE,
      summary: opts?.summary || null,
      payload: redactPayload(payload, opts?.redaction || REDACTION.NONE),
    };
    this.store.appendEvent(runId, event);
    return event;
  }
}
